File: LSB_Hide_Message.java
import java.io.*;
import java.lang.Object;
import javax.imageio.ImageIO;
import java.awt.image.*;
import java.util.*;

public class LSB_Hide_Message
{
	private String imagePath;
	private String imagename;
	private String imageType; // tested with bmp and png file
	private List<String> imageBytes;
	private List<String> dataBytes;
	private final int startIndex = 60;
	private final char stx = (char) 2;
	private final char etx = (char) 3;

	public LSB_Hide_Message(String imagePath) {
		this.imagePath = imagePath;
	}

	private void extractBytes() throws IOException {
		imageBytes = new ArrayList<String>();

		//Open the image
		File imgPath = new File(imagePath);
		BufferedImage imgBuffer = ImageIO.read(imgPath);

		imagename = imgPath.getName().substring(0, imgPath.getName().indexOf('.'));
		imageType = imgPath.getName().substring(imgPath.getName().indexOf('.') + 1, imgPath.getName().length());

		ByteArrayOutputStream b = new ByteArrayOutputStream();
		ImageIO.write(imgBuffer, "bmp", b);
		b.flush();
		byte[] imageData = b.toByteArray();
		b.close();


		for (int i = 0; i < imageData.length; i ++ ) {

			String s = String.format("%8s", Integer.toBinaryString(imageData[i] & 0xFF)).replace(' ', '0');

			imageBytes.add(s);
		}
	}

	private void getDataToHide() {
		dataBytes = new ArrayList<String>();

		String data = "" + stx;

		//get text data from user input
		System.out.print("Enter Secret Message : ");
		Scanner sc = new Scanner (System.in);
		data += sc.nextLine();
		data += etx;

		char [] charData = data.toCharArray();

		//convert data to binary
		for (int i = 0; i < charData.length; i++) {

			String s = String.format("%8s", Integer.toBinaryString(charData[i] & 0xFF)).replace(' ', '0');

			dataBytes.add(s);
		}
	}

	private boolean compressImage() {
		byte[] imageDataInDecimal = new byte[imageBytes.size()];
		for (int i = 0; i < imageBytes.size(); i++) {
			int imageData = Integer.parseInt(imageBytes.get(i), 2);

			imageDataInDecimal[i] = (byte)imageData;
		}

		try {
			// convert byte array back to BufferedImage
			BufferedImage bImageFromConvert = ImageIO.read(new ByteArrayInputStream(imageDataInDecimal));

			ImageIO.write(bImageFromConvert, imageType, new File(imagename+"-stego."+imageType));
		}
		catch (IOException ex) {
			System.out.println(ex.getMessage());
			return false;
		}

		return true;
	}

	private boolean canUsed(String binary) {
		return Integer.parseInt(binary, 2) == 255 || Integer.parseInt(binary, 2) == 254 ? false : true;
	}

	public boolean hide_message() {

		try {
			//get image data and convert to binary
			extractBytes();
		}
		catch(IOException ex) {
			return false;
		}

		//get data from user input and convert to binary
		getDataToHide();

		int counter = startIndex;
		for (int i = 0; i < dataBytes.size(); i ++) {

			char [] databit = dataBytes.get(i).toCharArray();
			for (int j = 0; j < databit.length; j ++) {
				String imageByte = imageBytes.get(counter).substring(0, imageBytes.get(i).length() - 1) + databit[j];

				if (canUsed(imageByte)) {
					imageBytes.set(counter, imageByte);
 					counter++;
				}
				else {
					j--;
					counter++;
				}
			}
		}

		if (compressImage()) return true;
		return false;

	}

	public String getMessage(int startindex, int endindex) {
		int counter = 0;
		String resultInBinary = "";
		String result = "";
		boolean checkStx = true;
		for (int i = startindex; i < endindex; i++) {
			if (canUsed(imageBytes.get(i))) {
				if (counter == imageBytes.get(i).length()) {
					int resultInInt = Integer.parseInt(resultInBinary, 2);
					char charRes = (char) resultInInt;
					if (checkStx) {
						if (charRes != stx) return "";
						else checkStx = false;
					}
					else {
						if (charRes != etx) {
							result += charRes;
						}
						else {
							break;
						}
					}
					resultInBinary = "";
					counter = 0;
					i--;
				}
				else {
					resultInBinary += imageBytes.get(i).substring(imageBytes.get(i).length() - 1, imageBytes.get(i).length());
					counter ++;
				}
			}
			else {
				endindex ++;
			}
		}

		return result;
	}

	public String extract_message() {
		try{
			extractBytes();	
		}
		catch (IOException ex) {

		}
		
		int counter = 0;
		String resultInBinary = "";
		String result = "";
		boolean checkStx = true;
		for (int i = startIndex; i < imageBytes.size(); i++) {
			if (canUsed(imageBytes.get(i))) {
				if (counter == imageBytes.get(i).length()) {
					int resultInInt = Integer.parseInt(resultInBinary, 2);
					char charRes = (char) resultInInt;
					if (checkStx) {
						if (charRes != stx) return "";
						else checkStx = false;
					}
					else {
						if (charRes != etx) {
							result += charRes;
						}
						else {
							break;
						}
					}
					resultInBinary = "";
					counter = 0;
					i--;
				}
				else {
					resultInBinary += imageBytes.get(i).substring(imageBytes.get(i).length() - 1, imageBytes.get(i).length());
					counter ++;
				}
			}
		}

		return result;		
	}

	public static void main(String [] args)
	{
		Scanner sc = new Scanner(System.in);

		while(true) {
			System.out.println("Select Option");
			System.out.println("1. Create Hidden Message");
			System.out.println("2. Open Hidden Message");
			System.out.println("3. Exit");
			System.out.print("Your Choice : ");
			int select = sc.nextInt();

			switch (select) {
				case 1 :
				{
					System.out.print("Open image : ");
					String filename = sc.next();
					LSB_Hide_Message ez1 = new LSB_Hide_Message(filename);
					System.out.println("\n=======================================================");
					if (ez1.hide_message()) System.out.println("Secret image created successfully");
					else System.out.println("Failed to create secret image");
				} break;

				case 2 :
				{
					System.out.print("Open secret image : ");
					String filename = sc.next();
					LSB_Hide_Message ez2 = new LSB_Hide_Message(filename);
					String message = ez2.extract_message();

					if (message != "") {
						System.out.println("\n=======================================================");
						System.out.println("Secret Message : " + message);
					}
					else {
						System.out.println("\n=======================================================");
						System.out.println("This image is not contain any message");
					} 
				} break;

				case 3:
				{
					System.exit(0);
				}

				default: 
				{
					System.out.println("\n=======================================================");
					System.out.println("Wrong input!");
				}
			}

			System.out.println("=======================================================\n");
		}
		
		
	}
}

File: LSB_Hide_Message.py
# from bitstring import BitArray
import binascii
import io
import sys
import readline
readline.parse_and_bind("tab: complete")

class LSB_Hide_Message:

	def __init__(self, imagePath):
		self.imagePath = imagePath
		self.imageName = imagePath[:imagePath.index('.')]
		self.imageBin = []
		self.messageBin = []
		self.STX = "\x02" 		# the message start with STX to check if the image has secret message 
		self.ETX = "\x03"		# ETX used to sign the end of message, so the iteration will stop
		self.startIndex = 60	# Insertion start from index 60, because if you insert from 0, the image will corrupt 

	def extractBytes(self):
		# open image and save image as binary
		with open(self.imagePath, "rb") as imageFile:
			f = imageFile.read()
			b = bytearray(f)

			imageFile.close()

			for byte in b:
				self.imageBin.append(bin(byte)[2:].zfill(8))


	def getHiddenData(self):

		print("Enter Secret Message : ")
		message = self.STX
		message = message + raw_input()
		message = message + self.ETX

		messageByte = bytearray(message.encode('UTF-8'))	# convert image to binary

		for char in messageByte:
			self.messageBin.append(bin(char)[2:].zfill(8))


	def compressImage(self):

		imageInDecimal = []

		for byte in self.imageBin:
			imageInDecimal.append(int(byte, 2))

		try:

			data = bytearray(imageInDecimal)
			with open(self.imageName + "-stego.bmp", "wb") as out_file:
				out_file.write(data)
				out_file.close()

			return True
		except Exception, e:
			print(str(e))
			return False

	def canUsed(self, bin):
		return int(bin, 2) <= 254	# this program use 8 bit data, so if there are negative number, it will convert to positive
									# such as if you have -1 then it will be 255. It will broke the image.

	def hide_message(self):
		try:
			self.extractBytes()
		except:
			return False

		self.getHiddenData()

		counter = self.startIndex

		# insert all message character to 1 digit image's LSB.
		# you can use 2 digits or 3 digits, but it will make the big differences with original image.
		for byte in self.messageBin:

			bitArray = str(byte.encode('UTF-8'))

			i = 0
			while i < len(bitArray):

				if self.canUsed(self.imageBin[counter]):

					self.imageBin[counter] = self.imageBin[counter][:1] + str(bitArray[i])	# insert bit per bit of message char to image by replacing
																							# the LSB digit

					i += 1

				counter += 1

		if self.compressImage():
			return True
		return False

	def extract_message(self):
		try:
			self.extractBytes()
		except:
			return False

		counter = 0
		checkStx = True
		resultInBinary = ""
		result = ""
		i = self.startIndex

		while i < len(self.imageBin):
			if self.canUsed(self.imageBin[i]):
				if counter == len(self.imageBin[i]):
					resultInChar = chr(int(resultInBinary, 2))

					if checkStx:
						if resultInChar != self.STX:
							return False

						checkStx = False

					else:
						if resultInChar != self.ETX:
							result = result + resultInChar
						else :
							return result;

					resultInBinary = ""
					counter = 0
					i -= 1

				else :
					resultInBinary = resultInBinary + self.imageBin[i][-1]
					counter += 1
			
			i += 1


if __name__ == '__main__':

	while True:
		print("Select option")
		print("1. Create hidden message")
		print("2. Open hidden message")
		print("3. Exit")
		sys.stdout.write("Your choice : ")
		choice = input()

		if choice == 1 :
			sys.stdout.write("Open image (only bmp image can be used) : ")
			filename = raw_input()
			lsb = LSB_Hide_Message(filename)
			if lsb.hide_message():
				print("\n=======================================================")
				print("Secret image created successfully")
			else:
				print("Failed to create secret image")
		elif choice == 2 :
			sys.stdout.write("Open secret image (only bmp image can be used) : ")
			filename = raw_input()
			lsb = LSB_Hide_Message(filename)
			message = lsb.extract_message()

			if not(message):
				print("\n=======================================================")
				print("This message is not contain any message")
			else :
				print("\n=======================================================")
				new_msg = ""
				counter1 = 0
				counter2 = 46

				for i in range(0, len(message)/46 + 1):
					new_msg = new_msg + message[counter1:counter2] + "\n"
					counter1 += 46
					counter2 += 46

				print("Secret message : \n")
				print(new_msg)

		elif choice == 3 :
			sys.exit()

		print("=======================================================\n")
